subgraph/schema.graphql

```diff
@@ -0,0 +1,100 @@
+type MACI @entity {
+  id: ID!
+  owner: Bytes! # address
+  stateTreeDepth: Int! # uint8
+  updatedAt: BigInt!
+  blockNumber: BigInt!
+  txHash: Bytes!
+
+  "state"
+  numRegistrar: BigInt!
+  numPoll: BigInt!
+  numVotes: BigInt!
+}
+
+type User @entity(immutable: true) {
+  id: Bytes! # address
+  timestamp: BigInt! # uint256
+  blockNumber: BigInt!
+  txHash: Bytes!
+
+  "relations"
+  polls: [Poll!]! @derivedFrom(field: "owner")
+  accounts: [Account!]! @derivedFrom(field: "owner")
+}
+
+type Account @entity {
+  id: ID! # pubkey
+  timestamp: BigInt! # uint256
+  blockNumber: BigInt!
+  txHash: Bytes!
+
+  "relations"
+  owner: User
+  polls: [Poll!]! @derivedFrom(field: "coordinator")
+  stateLeafs: [StateLeaf!]! @derivedFrom(field: "account")
+}
+
+type StateLeaf @entity {
+  id: ID! # stateIndex
+  voiceCreditBalance: BigInt! # uint256
+  timestamp: BigInt! # uint256
+  blockNumber: BigInt!
+  txHash: Bytes!
+
+  "relations"
+  account: Account!
+  topupHistory: [TopupCredit!]! @derivedFrom(field: "stateLeaf")
+}
+
+type Poll @entity {
+  id: Bytes! # poll address
+  pollId: BigInt! # uint256
+  duration: BigInt! # uint256
+  treeDepth: Int! # uint8
+  maxVoteOptions: BigInt!
+  messageProcessor: Bytes! # address
+  tally: Bytes! # address
+  subsidy: Bytes! # address
+  "merge state after ended"
+  stateRoot: BigInt # uint256
+  numSignups: BigInt! # uint256
```

Comment:
I believe `numSignups` should be on `MACI` not `Poll`

Reply:
so `numMessages` is number of votes, also including topup messages. is that correct?

Reply:
regarding the `topup` process, we need which `stateIndex` and how much `voteCredit` to be added to the right state leaf in the tree. but both required params is only available via function call, which need to handle seperately from the event handler, somehow network is not support the tracing API.

Reply:
> w need



> ~I believe `numSignups` should be on `MACI` not `Poll`~ - nevermind, I guess `numSignups` is also on Poll. So @ctrlc03 we now need on `numSignups` on Poll (#1101) in addition to on MACI?

totally missed this. `numSignups` on poll is the number of signups before the poll was "completed" (stateTree and message tree merged)

Reply:
> regarding the `topup` process, we need which `stateIndex` and how much `voteCredit` to be added to the right state leaf in the tree. but both required params is only available via function call, which need to handle seperately from the event handler, somehow network is not support the tracing API.

you have the data here:

https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/contracts/Poll.sol#L173 

The topup message will be an object:

{

msgType: 2
data: [stateIndex, amount, 0, 0, 0, ...]
}

Reply:
> > w need
> 
> > ~I believe `numSignups` should be on `MACI` not `Poll`~ - nevermind, I guess `numSignups` is also on Poll. So @ctrlc03 we now need on `numSignups` on Poll (#1101) in addition to on MACI?
> 
> totally missed this. `numSignups` on poll is the number of signups before the poll was "completed" (stateTree and message tree merged)

I'm not quite sure on this, since the users sign up to MACI without specifying to any certain poll. how can we count the number of signup in the certain poll? should we count when user publish message?

Reply:
> > regarding the `topup` process, we need which `stateIndex` and how much `voteCredit` to be added to the right state leaf in the tree. but both required params is only available via function call, which need to handle seperately from the event handler, somehow network is not support the tracing API.
> 
> you have the data here:
> 
> https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/contracts/Poll.sol#L173
> 
> The topup message will be an object:
> 
> {
> 
> msgType: 2 data: [stateIndex, amount, 0, 0, 0, ...] }

awesome! I'll update the subgraph based on this

Reply:
> > > w need
> > 
> > 
> > > ~I believe `numSignups` should be on `MACI` not `Poll`~ - nevermind, I guess `numSignups` is also on Poll. So @ctrlc03 we now need on `numSignups` on Poll (#1101) in addition to on MACI?
> > 
> > 
> > totally missed this. `numSignups` on poll is the number of signups before the poll was "completed" (stateTree and message tree merged)
> 
> I'm not quite sure on this, since the users sign up to MACI without specifying to any certain poll. how can we count the number of signup in the certain poll? should we count when user publish message?

So you can think of the number of signups to a Poll, the number of users that are signed up up until the end of the poll. So for instance we start 1 poll and 50 users sign up, then we complete it (50 signups there). then we start another poll and 10 more users sign up, then it completed, and we have 60 signups. Does that make sense?

Reply:
> > regarding the `topup` process, we need which `stateIndex` and how much `voteCredit` to be added to the right state leaf in the tree. but both required params is only available via function call, which need to handle seperately from the event handler, somehow network is not support the tracing API.
> 
> you have the data here:
> 
> https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/contracts/Poll.sol#L173
> 
> The topup message will be an object:
> 
> {
> 
> msgType: 2 data: [stateIndex, amount, 0, 0, 0, ...] }

thank you for give me some clue, I've updated

anyway, how can I get `stateIndex` and `voteWeight` from the emitted event `PublishMessage` since the message got hashed and encrypted before submitting on-chain?

Reply:
the reason I want to do this, because I would like to subtract the `voteCredits` from the `voteBalance` state in the certain account linked to the `stateIndex`? or I don’t need to do that, should I store the encrypted data as it is?

Reply:
> from the emitted event `PublishMessage` since the message got hashed and encrypted before submitting on-ch

you're not supposed to get that data, messages are private so they are encrypted before being sent to prevent its content from being seen 🙂

Reply:
> the encry

it'll be fine to store the encrypted data only, thank you!

Reply:
ahh I see, thank you!